### DISCUSSION TITLE
how to filter out the itemtype properties information from multiple element properties

### DISCUSSION THREAD

**Discussion by ghost** [iTwin/itwinjs-core#6620] [General] [answered]
After obtaining the property information of all graphical elements using `PresentationManager.getElementProperties`, how can I filter out the itemtype properties information? The material table【材料表】 in the image is a custom itemtype property table that I created.
Thanks for everyone's help.
![itemtypeProperty](https://github.com/iTwin/itwinjs-core/assets/34014227/ed363553-7dd1-4c4a-878e-520f7652edf1)

**Comment by grigasp** [comment, marked as answer]
Filtering out the information after you get the result is:
1. inefficient,
2. problematic, since the result doesn't contain metadata, which makes it impossible to identify what to remove (unless there's something you can trust to always stay the same, e.g. the category name).

The most efficient way would be to tell the Presentation system to not load those properties at all. However, the intention of `getElementProperties` API was to make it as simple as possible to retrieve properties displayed in the property grid, so it doesn't have this customization option. For that you would need to:

1. Find out where the properties come from. The [BisCore supplemental ruleset](https://github.com/iTwin/itwinjs-core/blob/master/presentation/backend/src/presentation-backend/supplemental-presentation-rules/BisCore.PresentationRuleSet.json) defines a number of `ContentModifier` rules with `relatedProperties` specifications that the Presentation system uses to find additional elements' properties. Your properties must be coming through one of these specifications and you must identify which one, based on relationship between the primary element's class and the properties' class.

2. Use the `getContent` API with a custom presentation ruleset. The custom ruleset should contain a similar `ContentModifier` rule, with similar `relatedProperties` specification, except that the `properties` attribute should be set to `_none_` (see [here](https://www.itwinjs.org/presentation/content/relatedpropertiesspecification/#attribute-properties)). This will basically disable the default supplemental rule.

3. Parse the result of `getContent` call into a format you want.

**Reply by ghost** [reply]
Thank you for your response. If I only want to retrieve the following properties information for all elements from a BIM file:
1. Element IDs, including the elementId from the original DGN file before it was published to the BIM file, and the instanceId in the BIM file;
2. ItemType properties information.

How should I write this presentation ruleset? Are there any similar  ruleset file available, or could you provide one? Thanks again.

**Reply by grigasp** [reply]
First, you have to know where the properties are located. From your list I know 2 of them:
- "instanceId in the BIM file" - `ECInstanceId` of the element you request properties for,
- "elementId from the original DGN file before it was published to the BIM file" - `Element -> ElementOwnsMultiAspects -> ExternalSourceAspect`, the property `ExternalSourceAspect.Identifier`.

I'm not sure what "ItemType properties information" is or how to find it relative to an element.

With the above information, you need the most efficient way to retrieve the information for your use case. The Presentation APIs are good when you want the properties to be formatted and ready to display to the end user. When loading information like IDs, where formatting is not involved, I suggest using plain [ECSQL queries](https://www.itwinjs.org/learning/ecsql/).

**Reply by ghost** [reply]
@grigasp The itemtype properties table was created by me using the Itemtype tool in MicroStation software, and I attached the relevant properties to the test elements within MicroStation. 

**here is the test bim file uploaded：**
[57a1bc88-e624-4f13-94d8-547fb0cf6a3f.zip](https://github.com/iTwin/itwinjs-core/files/14978566/57a1bc88-e624-4f13-94d8-547fb0cf6a3f.zip)

Taking the BIM file I uploaded as an example:
1. the itemtype properties exist in the bis_ElementMultiAspect table of the BIM file, with three elements (0x427, 0x437, 0x442) having this property information. 
2. The category name of the properties table is located in the ec_Class table, with an ID value of 338 and the Name as __x6750____x6599____x8868__ElementAspect. 
3. The property names are located in the ec_Property table, as shown in the three figures below.

**How should I write this presentation ruleset?** thanks again.

![01](https://github.com/iTwin/itwinjs-core/assets/34014227/1600ae82-7dac-4589-b62c-0d72ad444fb5)

![02](https://github.com/iTwin/itwinjs-core/assets/34014227/924133cb-f5e2-4f9a-90bb-08faa76c64b0)

![03](https://github.com/iTwin/itwinjs-core/assets/34014227/b2118ff7-dc7e-49fe-b981-db151ac4fa46)

**Reply by grigasp** [reply]
All 3 of them are strings and most likely don't need extra formatting. How likely is it for properties of different types to be attached to these elements? If unlikely, the easiest and most efficient way to for you to get the information is like this:
```ts
const ecsql = `
  SELECT e.ECInstanceId, external_source.Identifier, item_type.*
  FROM bis.Element e
  LEFT JOIN bis.ExternalSourceAspect AS external_source ON e.ECInstanceId = external_source.Element.Id
  LEFT JOIN DgnCustomItemTypes___x91D1____x7ED3____x5C5E____x6027____x5E93__.__x6750____x6599____x8868__ElementAspect AS item_type ON e.ECInstanceId = item_type.Element.Id
`;
for await (const row of imodel.createQueryReader(ecsql)) {
  console.log(row);
}
```
